### PR TITLE
Implement tracker blacklist

### PR DESCRIPTION
**Pull request by nitros12**
Resolves: #169 

Adds the commands: `trackblacklistadd` and `trackblacklistremove`.

Adds a new enum type:
```sql
CREATE TYPE trackingblacklisttype AS ENUM (
    'USER',
    'CHAN'
);
```
Adds a new table:
```sql
CREATE TABLE tracking_blacklists (
    guild_id bigint NOT NULL,
    type trackingblacklisttype NOT NULL,
    data_id bigint NOT NULL
);

ALTER TABLE ONLY tracking_blacklists
    ADD CONSTRAINT tracking_blacklists_pkey PRIMARY KEY (guild_id, type, data_id);

ALTER TABLE ONLY tracking_blacklists
    ADD CONSTRAINT tracking_blacklists_guild_id_fkey FOREIGN KEY (guild_id) REFERENCES guilds(guild_id);
```

### CODE REVIEW

File: futaba/cogs/settings/core.py [hunk 1]
```diff
@@ -342,3 +343,55 @@ def mention(role):
 
         await ctx.send(embed=embed)
         self.journal.send("roles/jail", ctx.guild, content, icon="settings", role=role)
+
+    @commands.command(name="trackblacklistadd")
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: There should be at least one alias that shortens this command. Given that this follows the add/remove pattern, I think it's best split out into a command group ("trackerblacklist") with subcommands ("add", "remove", "show"...) (see `!role`)

File: futaba/cogs/settings/core.py [hunk 2]
```diff
@@ -342,3 +343,55 @@ def mention(role):
...
+    @commands.guild_only()
+    @permissions.check_mod()
+    async def tracker_blacklist_add(
+        self, ctx, *, user_or_channel: Union[MemberConv, TextChannelConv]
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: :+1:

File: futaba/cogs/settings/core.py [hunk 3]
```diff
@@ -342,3 +343,55 @@ def mention(role):
...
+    ):
+        """ Add a user or channel to the tracking blacklist. """
+
+        logger.info(
+            "Adding '%s' (%d) to the tracking blacklist for guild '%s' (%d)",
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: Maybe add a `"user" if isinstance(user_or_channel, discord.Member) else "channel"` to the log.
E.g. "Adding user 'aismallard' (111111111111) to the tracking blacklist for guild ..."

File: futaba/cogs/settings/core.py [hunk 4]
```diff
@@ -342,3 +343,55 @@ def mention(role):
...
+            user_or_channel.name,
+            user_or_channel.id,
+            ctx.guild.name,
+            ctx.guild.id,
+        )
+
+        with self.bot.sql.transaction():
+            self.bot.sql.settings.add_to_tracking_blacklist(ctx.guild, user_or_channel)
+
+        embed = discord.Embed(colour=discord.Colour.green())
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: Should be `dark_teal()`. See [here](https://github.com/strinking/futaba/blob/master/CONTRIBUTING.md).

File: futaba/cogs/settings/core.py [hunk 5]
```diff
@@ -342,3 +343,55 @@ def mention(role):
...
+        logger.info(
+            "Adding '%s' (%d) to the tracking blacklist for guild '%s' (%d)",
+            user_or_channel.name,
+            user_or_channel.id,
+            ctx.guild.name,
+            ctx.guild.id,
+        )
+
+        with self.bot.sql.transaction():
+            self.bot.sql.settings.add_to_tracking_blacklist(ctx.guild, user_or_channel)
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: We might want to consider having a model just for tracking settings.

File: futaba/cogs/settings/core.py [hunk 6]
```diff
@@ -342,3 +343,55 @@ def mention(role):
...
+
+        embed = discord.Embed(colour=discord.Colour.green())
+        embed.description = f"Added {user_or_channel.mention} to the tracking blacklist"
+
+        await ctx.send(embed=embed)
+
+    @commands.command(name="trackblacklistremove")
+    @commands.guild_only()
+    @permissions.check_mod()
+    async def tracker_blacklist_remove(
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: ditto on aliases and command group

File: futaba/cogs/settings/core.py [hunk 7]
```diff
@@ -342,3 +343,55 @@ def mention(role):
...
+        self, ctx, *, user_or_channel: Union[MemberConv, TextChannelConv]
+    ):
+        """ Remove a user or channel from the tracking blacklist. """
+
+        logger.info(
+            "Removing '%s' (%d) from the tracking blacklist for guild '%s' (%d)",
+            user_or_channel.name,
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: ditto

File: futaba/cogs/settings/core.py [hunk 8]
```diff
@@ -342,3 +343,55 @@ def mention(role):
...
+            ctx.guild.name,
+            ctx.guild.id,
+        )
+
+        with self.bot.sql.transaction():
+            self.bot.sql.settings.remove_from_tracking_blacklist(
+                ctx.guild, user_or_channel
+            )
+
+        embed = discord.Embed(colour=discord.Colour.green())
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: ditto

File: futaba/cogs/tracker/core.py [hunk 9]
```diff
@@ -468,6 +557,17 @@ def build_embed(message):
...
+        if blacklist.is_blocked(member):
+            logger.debug(
+                "Ignoring member %s (%d) leaving guild '%s' (%d) due to the user being blacklisted",
+                member.name,
+                member.id,
+                member.guild.name,
+                member.guild.id,
+            )
+            return
+
```
Comment (ammongit) [inline futaba/cogs/tracker/core.py]: :+1:

Comment (ammongit) [reply futaba/cogs/tracker/core.py]: (Though blacklisting join/leave events is definitely an interesting situation. Wonder if there's a good practical usecase for it)

File: futaba/sql/models/settings.py [hunk 10]
```diff
@@ -17,12 +17,15 @@
...
 import functools
 import logging
 
 import discord
-from sqlalchemy import BigInteger, Boolean, Column, Table, Unicode
-from sqlalchemy import CheckConstraint, ForeignKey, SmallInteger
+
+from futaba.utils import partition_on
+from sqlalchemy import (BigInteger, Boolean, CheckConstraint, Column, Enum,
+                        ForeignKey, SmallInteger, Table, Unicode)
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: `futaba.*` imports should be treated like they are relative, and go after third-party imports.
e.g.
```py
import discord
import toml

from futaba.utils import escape_backticks
from .local import helper_func
```

File: futaba/sql/models/settings.py [hunk 11]
```diff
@@ -94,13 +97,56 @@ def __iter__(self):
...
+        blacklisted_channels, blacklisted_users = partition_on(
+            lambda block: block[0] is TrackingBlacklistType.CHAN,
+            blacklist,
+            lambda block: block[1],
+        )
+
+        self.blacklisted_channels, self.blacklisted_users = (
+            set(blacklisted_channels),
+            set(blacklisted_users),
+        )
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: :+1:

File: futaba/sql/models/settings.py [hunk 12]
```diff
@@ -94,13 +97,56 @@ def __iter__(self):
...
+
+    def is_blocked(self, user_or_channel):
+        if isinstance(user_or_channel, (discord.User, discord.Member)):
```
Comment (ammongit) [inline futaba/sql/models/settings.py]:
```suggestion
        if isinstance(user_or_channel, discord.abc.User):
```

File: futaba/sql/models/settings.py [hunk 13]
```diff
@@ -94,13 +97,56 @@ def __iter__(self):
...
+    def remove_block(self, user_or_channel):
+        if isinstance(user_or_channel, (discord.User, discord.Member)):
+            self.blacklisted_users.discard(user_or_channel.id)
+        else:
+            self.blacklisted_channels.discard(user_or_channel.id)
+
+
+class TrackingBlacklistType(enum.Enum):
+    USER = 1
+    CHAN = 2
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: Don't we already have `LocationType`? Could that be reused here?

File: futaba/sql/models/settings.py [hunk 14]
```diff
@@ -156,8 +202,18 @@ def __init__(self, sql, meta):
                 name="special_role_jail_not_member_check",
             ),
         )
+        self.tb_tracking_blacklists = Table(
+            "tracking_blacklists",
+            meta,
+            Column(
+                "guild_id", BigInteger, ForeignKey("guilds.guild_id"), primary_key=True
+            ),
+            Column("type", Enum(TrackingBlacklistType), primary_key=True),
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: Is there a difference between `primary_key`ing all columns vs just adding a `UniqueConstraint`? imo the latter looks better, but I don't think it matters.

File: futaba/sql/models/settings.py [hunk 15]
```diff
@@ -348,3 +404,66 @@ def set_special_roles(self, guild, **attrs):
...
+            "Adding '%s' (%d) to the tracking blacklist for guild '%s' (%d)",
+            user_or_channel.name,
+            user_or_channel.id,
+            guild.name,
+            guild.id,
+        )
+
+        block_type = (
+            TrackingBlacklistType.USER
+            if isinstance(user_or_channel, (discord.User, discord.Member))
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: ditto (re: `discord.abc.User`)

File: futaba/sql/models/settings.py [hunk 16]
```diff
@@ -348,3 +404,66 @@ def set_special_roles(self, guild, **attrs):
...
+            "Deleting '%s' (%d) from the tracking blacklist for guild '%s' (%d)",
+            user_or_channel.name,
+            user_or_channel.id,
+            guild.name,
+            guild.id,
+        )
+
+        block_type = (
+            TrackingBlacklistType.USER
+            if isinstance(user_or_channel, (discord.User, discord.Member))
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: ditto

File: futaba/utils.py [hunk 17]
```diff
@@ -265,3 +265,24 @@ def escape_backticks(content):
...
+            else:
+                right.append(i)
+    else:
+        for i in iterable:
+            if predicate(i):
+                left.append(map_fn(i))
+            else:
+                right.append(map_fn(i))
+
+    return left, right
```
Comment (ammongit) [inline futaba/utils.py]: :+1:

**Comment by ammongit** [comment]
We don't need to manually run SQL to add types or tables since SQLAlchemy does that for us.

File: futaba/cogs/settings/core.py (hunk 1 above)
Comment (nitros12) [reply futaba/cogs/settings/core.py]: yup, should viewing the blacklist be mod only?

Comment (ammongit) [reply futaba/cogs/settings/core.py]: I think that makes sense. You still need to apply `@permissions.check_mod()` to each of the subcommands though I believe.

File: futaba/sql/models/settings.py (hunk 14 above)
Comment (nitros12) [reply futaba/sql/models/settings.py]: afaik the only difference is that using multiple primary_key's results in a ddl for `ADD CONSTRAINT ... PRIMARY KEY ...;` whereas a unique constraint generates `ADD CONSTRAINT ... UNIQUE CONSTRAINT ...;` in postgres there's no difference as PRIMARY KEY is shorthand for UNIQUE NOT NULL.

File: futaba/cogs/settings/core.py (hunk 1 above)
Comment (Ma-wa-re) [reply futaba/cogs/settings/core.py]: You don't if the group command is mod only

File: futaba/cogs/settings/core.py [hunk 18]
```diff
@@ -342,3 +345,104 @@ def mention(role):
...
+                channel = discord.utils.get(ctx.guild.channels, id=channel_id)
+                channel_msg.write(channel.mention)
+
+            embed.add_field(name="Blacklisted channels", value=channel_msg)
+
+        if blacklist.blacklisted_users:
+            user_msg = StringBuilder(sep=", ")
+            for user_id in blacklist.blacklisted_users:
+                user = discord.utils.get(
+                    chain(ctx.guild.members, ctx.bot.users), id=user_id
```
Comment (ammongit) [inline futaba/cogs/settings/core.py]: :+1:

File: futaba/sql/models/settings.py [hunk 19]
```diff
@@ -17,14 +17,28 @@
...
+    ForeignKey,
+    SmallInteger,
+    Table,
+    Unicode,
+    UniqueConstraint,
+)
 from sqlalchemy.sql import select
 
+from ...enums import LocationType
+from ...utils import partition_on
```
Comment (ammongit) [inline futaba/sql/models/settings.py]:
```suggestion
from futaba.utils import partition_on
```

File: futaba/sql/models/settings.py [hunk 20]
```diff
@@ -17,14 +17,28 @@
...
+    Enum,
+    ForeignKey,
+    SmallInteger,
+    Table,
+    Unicode,
+    UniqueConstraint,
+)
 from sqlalchemy.sql import select
 
+from ...enums import LocationType
```
Comment (ammongit) [inline futaba/sql/models/settings.py]:
```suggestion
from futaba.enums import LocationType
```

File: futaba/sql/models/settings.py [hunk 21]
```diff
@@ -156,8 +208,17 @@ def __init__(self, sql, meta):
                 name="special_role_jail_not_member_check",
             ),
         )
+        self.tb_tracking_blacklists = Table(
+            "tracking_blacklists",
+            meta,
+            Column("guild_id", BigInteger, ForeignKey("guilds.guild_id"), index=True),
+            Column("type", Enum(LocationType)),
+            Column("data_id", BigInteger),
+            UniqueConstraint("guild_id", "type", "data_id"),
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: Add a `name`:
```suggestion
            UniqueConstraint("guild_id", "type", "data_id", name="tracking_blacklist_uq"),
```

File: futaba/sql/models/settings.py [hunk 22]
```diff
@@ -156,8 +208,17 @@ def __init__(self, sql, meta):
...
+        )
```
Comment (ammongit) [inline futaba/sql/models/settings.py]: Since we're using `LocationType` now, consider adding a `CheckConstraint()` to make sure it's only channel or user.

**Pull request merged**